### ISSUE TITLE
Current ResultAction design is bad

### ISSUE THREAD

**Issue by Makman2** [coala-analyzer/coala#1168]
To run an action it is necessary to instantiate it, but with the current model this is totally superfluos. Nearly everything happens statically, but not the applying.
Imo there are two options:
- Make `apply` and `apply_from_section` static
- Create a uniform constructor that takes a result (and maybe file-dicts). Make every method that needs to operate on the result (like `is_applicable`) non-static.

I would prefer the second one, constructs like this make sense (at least for me):
```python
action = MyAction(result, maybe_additional_parameters)
action.is_applicable()
# Hopefully yields True

action.apply()
```
This needs some concepts for auto-applyable functions, but this should be a nice and intuitive way for handling actions^^

**Comment by Makman2** [comment]
For default actions we should maybe allow such constructs if actions need parameters:
```
default_actions = SpaceConsistencyBear: XBear(1, 2, "arg")
```

**Comment by sils1297** [comment]
@Makman2 I think it would be simpler just making everything static.

**Comment by Uran198** [comment]
I also think that making it non-static is better and more intuitive idea. Also it will decrease redundancy in interfaces of functions. Instead of passing arguments to every static function we will pass it only to the constructor.
Also, I'd work on this, if there is no objection on making it non-static :smile:

**Comment by sils1297** [comment]
Wouldn't it be way easier to just make a decorator so if you want to write an action (that does something, like a function, not *is* something like a class) you just write a function with your preferred name and the decorator would take care of the rest?

**Comment by Uran198** [comment]
@sils1297 I'm probably missing something, but an action usually overrides not only `apply`, but also `is_applicable` method, so how would you use decorator instead of it?

**Comment by sils1297** [comment]
Well for many actions `is_applicable` isn't relevant, for those where it is, the method could be given as an argument to the decorator:

```python
@resultaction(is_applicable=lambda result: result.debug_msg is not None)
def print_debug_message(result):
    print(result.debug_msg)
```

**Comment by sils1297** [comment]
the decorator could btw. easily generate a class like we have so we would be backwards compatible and if one likes the class better he can just use that

**Comment by Uran198** [comment]
So, this decorator will instantiate `ResultAction` and return it instead of a function with some substituted method? It just may seem strange if someone would want also to override `apply_from_section` and his functions may be rather large. It would be like he is creating a new class, but very implicitly.

**Comment by sils1297** [comment]
if you do a complicated thing you can still use the class, but in the simple case you could use the decorator

**Comment by Uran198** [comment]
@sils1297 Yes, the decorator would be a good idea.
I think we may a little off topic here, the original proposal was to either make all static or all non-static in `ResultAction`. I like non-static idea as it will decrease redundancy.

**Comment by sils1297** [comment]
It's kinda related I think... guess we want to have one issue making it nonstatic and another one introducing the decorator? For the nonstatic thing it's important that custom parameters are still parameters of the apply method because they might change after instanciation

**Comment by Makman2** [comment]
The decorator is kind of a feature for the new ResultAction design. The new class design is anyway needed^^ --> So yes, file a new issue that is blocked by/dependent on this one :+1:

**Comment by Uran198** [comment]
The decorated function will have to have those parameters, but not necessary apply method, if we'll substitute a class or class instance for a function in decorator.
And even the decorated function will be able to live without them and decorator could use `types.MethodType` to bound it to the instance, but in this case code would be kind of strange.
